File: roadmap-checkpoints/CassaV6-roadmap-checkpoint-P2B-AUTH-STATUS-20260901/ROADMAP_V6/BLUETOOTH/CASSAV6_BLUETOOTH_PROTOCOL_20260820/raspberry/test/reliable-channel.test.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  RELIABLE_FRAME_TYPES
} from "../dist/protocol/FrameCodec.js";
import {
  InMemoryReliableChannelStoreV1,
  ReliableChannelError,
  ReliableChannelV1
} from "../dist/protocol/ReliableChannel.js";

const KEY_A_TO_B = Buffer.alloc(32, 0x41);
const KEY_B_TO_A = Buffer.alloc(32, 0x42);
const PREFIX_A_TO_B = Buffer.from("0102030405060708", "hex");
const PREFIX_B_TO_A = Buffer.from("1112131415161718", "hex");
const START = 1_800_000_000_000;
const PEER_TRUST_A = "a".repeat(64);
const PEER_TRUST_B = "b".repeat(64);
const PEER_TRUST_C = "c".repeat(64);

function createPair(options = {}) {
  let now = START;
  const aToB = [];
  const bToA = [];
  const deliveredA = [];
  const deliveredB = [];
  const storeA = options.storeA ?? new InMemoryReliableChannelStoreV1();
  const storeB = options.storeB ?? new InMemoryReliableChannelStoreV1();
  const a = new ReliableChannelV1({
    transport: {
      async send(frame) {
        options.onSendA?.(frame, storeA);
        aToB.push(Buffer.from(frame));
      }
    },
    store: storeA,
    peerTrustId: options.peerTrustIdA ?? PEER_TRUST_B,
    mtu: options.mtu ?? 23,
    txKey: KEY_A_TO_B,
    rxKey: KEY_B_TO_A,
    txNoncePrefix: PREFIX_A_TO_B,
    rxNoncePrefix: PREFIX_B_TO_A,
    onMessage: async (message) => {
      deliveredA.push({
        type: message.type,
        payload: Buffer.from(message.payload)
      });
    },
    maxAttempts: options.maxAttempts ?? 3,
    baseRetryMs: options.baseRetryMs ?? 100,
    maxRetryMs: options.maxRetryMs ?? 1_000,
    random: () => 0,
    now: () => now
  });
  const b = new ReliableChannelV1({
    transport: {
      async send(frame) {
        if (options.rejectSendB) throw new Error("ACK transport rejected");
        bToA.push(Buffer.from(frame));
      }
    },
    store: storeB,
    peerTrustId: options.peerTrustIdB ?? PEER_TRUST_A,
    mtu: options.mtu ?? 23,
    txKey: KEY_B_TO_A,
    rxKey: KEY_A_TO_B,
    txNoncePrefix: PREFIX_B_TO_A,
    rxNoncePrefix: PREFIX_A_TO_B,
    onMessage: async (message) => {
      if (options.rejectAtB) throw new Error("rejected by test handler");
      options.onMessageB?.(message);
      deliveredB.push({
        type: message.type,
        payload: Buffer.from(message.payload)
      });
    },
    maxAttempts: options.maxAttempts ?? 3,
    baseRetryMs: options.baseRetryMs ?? 100,
    maxRetryMs: options.maxRetryMs ?? 1_000,
    random: () => 0,
    now: () => now
  });
  return {
    a,
    b,
    aToB,
    bToA,
    deliveredA,
    deliveredB,
    storeA,
    storeB,
    setNow(value) {
      now = value;
    }
  };
}

async function deliver(queue, target) {
  const frames = queue.splice(0);
  let result = null;
  for (const frame of frames) result = await target.receiveFragment(frame);
  return { frames, result };
}

test("durable message commits before transport and clears only after ACK", async () => {
  let observedCommitted = false;
  const pair = createPair({
    onSendA(_frame, store) {
      observedCommitted ||= store.snapshot().outboxDepth === 1;
    }
  });
  const sent = await pair.a.send({
    type: RELIABLE_FRAME_TYPES.SHADOW_DIAGNOSTIC,
    payload: Buffer.from("health"),
    durable: true,
    messageId: "00112233445566778899aabbccddeeff"
  });
  assert.equal(sent.durableCommitted, true);
  assert.equal(observedCommitted, true);
  assert.equal(pair.storeA.snapshot().outboxDepth, 1);

  await deliver(pair.aToB, pair.b);
  assert.deepEqual(pair.deliveredB, [
    {
      type: RELIABLE_FRAME_TYPES.SHADOW_DIAGNOSTIC,
      payload: Buffer.from("health")
    }
  ]);
  assert.equal(pair.storeB.snapshot().inboxDedupDepth, 1);
  await deliver(pair.bToA, pair.a);
  assert.equal(pair.storeA.snapshot().outboxDepth, 0);
  assert.equal(pair.a.snapshot().pendingMessages, 0);
  assert.equal(pair.a.snapshot().acknowledgementsRx, 1);
  pair.a.close();
  pair.b.close();
});

test("duplicate delivery is suppressed but acknowledged again", async () => {
  const pair = createPair();
  await pair.a.send({
    type: RELIABLE_FRAME_TYPES.DATA,
    payload: Buffer.from("once"),
    messageId: "10112233445566778899aabbccddeeff"
  });
  const original = pair.aToB.map(Buffer.from);
  await deliver(pair.aToB, pair.b);
  const firstAckFrames = pair.bToA.length;
  for (const frame of original) pair.aToB.push(Buffer.from(frame));
  const replay = await deliver(pair.aToB, pair.b);
  assert.equal(replay.result.duplicate, true);
  assert.equal(pair.deliveredB.length, 1);
  assert.ok(pair.bToA.length > firstAckFrames);
  assert.equal(pair.b.snapshot().duplicates, 1);
  pair.a.close();
  pair.b.close();
});

test("retry backoff is deterministic and suspends after the configured limit", async () => {
  const pair = createPair({ maxAttempts: 2, baseRetryMs: 100 });
  await pair.a.send({
    type: RELIABLE_FRAME_TYPES.DATA,
    payload: Buffer.from("retry"),
    durable: true,
    messageId: "20112233445566778899aabbccddeeff"
  });
  const firstTransmissionFrames = pair.aToB.length;
  pair.setNow(START + 99);
  assert.deepEqual(await pair.a.tick(), {
    retried: 0,
    suspended: 0,
    expired: 0
  });
  pair.setNow(START + 100);
  assert.equal((await pair.a.tick()).retried, 1);
  assert.equal(pair.aToB.length, firstTransmissionFrames * 2);
  pair.setNow(START + 300);
  assert.equal((await pair.a.tick()).suspended, 1);
  assert.equal(pair.a.snapshot().suspendedMessages, 1);
  assert.equal(pair.storeA.snapshot().outboxDepth, 1);
  assert.equal(await pair.a.resumeSuspended(), 1);
  pair.a.close();
  pair.b.close();
});

test("durable outbox is restored into a fresh session without data loss", async () => {
  const store = new InMemoryReliableChannelStoreV1();
  const first = createPair({ storeA: store });
  await first.a.send({
    type: RELIABLE_FRAME_TYPES.ROUTE_ADVERTISEMENT,
    payload: Buffer.from("route"),
    durable: true,
    messageId: "30112233445566778899aabbccddeeff"
  });
  assert.equal(store.snapshot().outboxDepth, 1);
  first.a.close();
  first.b.close();

  const second = createPair({ storeA: store });
  assert.equal(await second.a.restoreDurableOutbox(), 1);
  await deliver(second.aToB, second.b);
  await deliver(second.bToA, second.a);
  assert.equal(store.snapshot().outboxDepth, 0);
  assert.deepEqual(second.deliveredB[0], {
    type: RELIABLE_FRAME_TYPES.ROUTE_ADVERTISEMENT,
    payload: Buffer.from("route")
  });
  second.a.close();
  second.b.close();
});

test("durable outbox and dedup never cross peer trust contexts", async () => {
  const store = new InMemoryReliableChannelStoreV1();
  const first = createPair({ storeA: store, peerTrustIdA: PEER_TRUST_B });
  await first.a.send({
    type: RELIABLE_FRAME_TYPES.DATA,
    payload: Buffer.from("peer-b"),
    durable: true,
    messageId: "31112233445566778899aabbccddeeff"
  });
  first.a.close();
  first.b.close();

  const wrongPeer = createPair({ storeA: store, peerTrustIdA: PEER_TRUST_C });
  assert.equal(await wrongPeer.a.restoreDurableOutbox(), 0);
  await wrongPeer.a.send({
    type: RELIABLE_FRAME_TYPES.DATA,
    payload: Buffer.from("peer-c"),
    durable: true,
    messageId: "31112233445566778899aabbccddeeff"
  });
  assert.equal(store.snapshot().outboxDepth, 2);
  wrongPeer.a.close();
  wrongPeer.b.close();

  const restored = createPair({ storeA: store, peerTrustIdA: PEER_TRUST_B });
  assert.equal(await restored.a.restoreDurableOutbox(), 1);
  restored.a.close();
  restored.b.close();
});

test("upper-layer failure is not acknowledged and remains retryable", async () => {
  const failing = createPair({ rejectAtB: true });
  await failing.a.send({
    type: RELIABLE_FRAME_TYPES.DATA,
    payload: Buffer.from("retry upper layer"),
    messageId: "40112233445566778899aabbccddeeff"
  });
  await assert.rejects(
    () => deliver(failing.aToB, failing.b),
    (error) =>
      error instanceof ReliableChannelError && error.code === "DELIVERY_FAILED"
  );
  assert.equal(failing.bToA.length, 0);
  assert.equal(failing.storeB.snapshot().inboxDedupDepth, 0);
  failing.a.close();
  failing.b.close();
});

test("expired durable messages are removed and never retried", async () => {
  const pair = createPair();
  await pair.a.send({
    type: RELIABLE_FRAME_TYPES.DATA,
    payload: Buffer.from("short lived"),
    durable: true,
    ttlMs: 1_000,
    messageId: "50112233445566778899aabbccddeeff"
  });
  pair.setNow(START + 1_000);
  assert.equal((await pair.a.tick()).expired, 1);
  assert.equal(pair.storeA.snapshot().outboxDepth, 0);
  assert.equal(pair.a.snapshot().pendingMessages, 0);
  pair.a.close();
  pair.b.close();
});

test("duplicate pending message ids fail closed without replacing key material", async () => {
  const pair = createPair();
  const messageId = "60112233445566778899aabbccddeeff";
  await pair.a.send({
    type: RELIABLE_FRAME_TYPES.DATA,
    payload: Buffer.from("first"),
    messageId
  });
  const framesBefore = pair.aToB.map(Buffer.from);

  await assert.rejects(
    () =>
      pair.a.send({
        type: RELIABLE_FRAME_TYPES.DATA,
        payload: Buffer.from("replacement"),
        messageId
      }),
    (error) =>
      error instanceof ReliableChannelError && error.code === "OUTBOX_CONFLICT"
  );
  assert.deepEqual(pair.aToB, framesBefore);
  assert.equal(pair.a.snapshot().pendingMessages, 1);
  pair.a.close();
  pair.b.close();
});

test("clock regression and user-created ACK fail closed", async () => {
  const pair = createPair();
  pair.setNow(START + 100);
  await pair.a.tick();
  pair.setNow(START + 99);
  await assert.rejects(
    () => pair.a.tick(),
    (error) =>
      error instanceof ReliableChannelError && error.code === "CLOCK_REGRESSION"
  );
  pair.setNow(START + 101);
  await assert.rejects(
    () =>
      pair.a.send({
        type: RELIABLE_FRAME_TYPES.ACK,
        payload: Buffer.alloc(16)
      }),
    (error) =>
      error instanceof ReliableChannelError && error.code === "ACK_RESERVED"
  );
  pair.a.close();
  pair.b.close();
});

test("close waits for the serialized operation before wiping channel state", async () => {
  let releaseTransport;
  let reportTransportStarted;
  const transportStarted = new Promise((resolve) => {
    reportTransportStarted = resolve;
  });
  const transportReleased = new Promise((resolve) => {
    releaseTransport = resolve;
  });
  const channel = new ReliableChannelV1({
    peerTrustId: PEER_TRUST_B,
    transport: {
      async send() {
        reportTransportStarted();
        await transportReleased;
      }
    },
    mtu: 23,
    txKey: KEY_A_TO_B,
    rxKey: KEY_B_TO_A,
    txNoncePrefix: PREFIX_A_TO_B,
    rxNoncePrefix: PREFIX_B_TO_A,
    onMessage() {},
    now: () => START,
    random: () => 0
  });
  const send = channel.send({
    type: RELIABLE_FRAME_TYPES.DATA,
    payload: Buffer.from("in flight"),
    messageId: "70112233445566778899aabbccddeeff"
  });
  await transportStarted;
  let closeCompleted = false;
  const close = channel.close().then(() => {
    closeCompleted = true;
  });
  await Promise.resolve();
  assert.equal(closeCompleted, false);
  releaseTransport();
  await send;
  await close;
  assert.equal(channel.snapshot().pendingMessages, 0);
  await assert.rejects(
    () =>
      channel.send({
        type: RELIABLE_FRAME_TYPES.DATA,
        payload: Buffer.alloc(0)
      }),
    (error) =>
      error instanceof ReliableChannelError && error.code === "CHANNEL_CLOSED"
  );
});

test("decoded payload is wiped when ACK transport fails", async () => {
  let deliveredPayload;
  const pair = createPair({
    rejectSendB: true,
    onMessageB(message) {
      deliveredPayload = message.payload;
    }
  });
  await pair.a.send({
    type: RELIABLE_FRAME_TYPES.DATA,
    payload: Buffer.from("wipe after ACK failure"),
    messageId: "80112233445566778899aabbccddeeff"
  });
  await assert.rejects(() => deliver(pair.aToB, pair.b), /ACK transport rejected/);
  assert.ok(deliveredPayload.every((value) => value === 0));
  await pair.a.close();
  await pair.b.close();
});

File: V5BT-sorgenti-e-roadmap-rimanente-20260818/APPLICATIVI/Palmare/android-app/app/src/test/java/com/sentrapa/cassav6/webkiosk/bluetooth/ReliableChannelTest.kt
package com.sentrapa.cassav6.webkiosk.bluetooth

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

class ReliableChannelTest {
    @Test
    fun `durable delivery ACK and persistent dedup complete end to end`() {
        val clock = MutableClock(NOW)
        val outboundA = mutableListOf<ByteArray>()
        val outboundB = mutableListOf<ByteArray>()
        val delivered = mutableListOf<ByteArray>()
        val storeA = InMemoryReliableChannelStoreV1()
        val storeB = InMemoryReliableChannelStoreV1()
        var committedBeforeTransport = false
        val channelA = channel(
            transport = {
                committedBeforeTransport =
                    committedBeforeTransport || storeA.snapshot().outboxDepth == 1
                outboundA += it.copyOf()
            },
            store = storeA,
            clock = clock,
            txKey = KEY_A,
            rxKey = KEY_B,
            txPrefix = PREFIX_A,
            rxPrefix = PREFIX_B,
            generatedIds = IdSource(ACK_A_ID)
        )
        val channelB = channel(
            transport = { outboundB += it.copyOf() },
            store = storeB,
            clock = clock,
            txKey = KEY_B,
            rxKey = KEY_A,
            txPrefix = PREFIX_B,
            rxPrefix = PREFIX_A,
            generatedIds = IdSource(ACK_B_ID, ACK_B_DUPLICATE_ID),
            peerTrustId = PEER_TRUST_A,
            onMessage = { delivered += it.payload.copyOf() }
        )

        try {
            val sent = channelA.send(
                ReliableChannelSendInputV1(
                    type = ReliableFrameTypeV1.SHADOW_DIAGNOSTIC,
                    payload = "health-only".toByteArray(),
                    durable = true,
                    messageId = MESSAGE_ID
                )
            )
            assertEquals(MESSAGE_ID, sent.messageId)
            assertTrue(sent.durableCommitted)
            assertTrue(committedBeforeTransport)
            assertEquals(1, storeA.snapshot().outboxDepth)
            val originalFrames = outboundA.map { it.copyOf() }

            val receiveResults = drain(outboundA, channelB)
            assertTrue(receiveResults.last().complete)
            assertTrue(receiveResults.last().delivered)
            assertArrayEquals("health-only".toByteArray(), delivered.single())
            assertEquals(1, storeB.snapshot().inboxDedupDepth)

            drain(outboundB, channelA)
            assertEquals(0, storeA.snapshot().outboxDepth)
            assertEquals(0, channelA.snapshot().pendingMessages)
            assertEquals(1, channelA.snapshot().acknowledgementsRx)

            val duplicateResults = originalFrames.map(channelB::receiveFragment)
            assertTrue(duplicateResults.last().duplicate)
            assertFalse(duplicateResults.last().delivered)
            assertEquals(1, channelB.snapshot().duplicates)
            assertEquals(1, delivered.size)
        } finally {
            channelA.close()
            channelB.close()
            storeA.close()
            storeB.close()
            outboundA.forEach { it.fill(0) }
            outboundB.forEach { it.fill(0) }
            delivered.forEach { it.fill(0) }
        }
    }

    @Test
    fun `durable outbox restores into a fresh channel and clears after ACK`() {
        val clock = MutableClock(NOW)
        val durableStore = InMemoryReliableChannelStoreV1()
        val abandonedFrames = mutableListOf<ByteArray>()
        val first = channel(
            transport = { abandonedFrames += it.copyOf() },
            store = durableStore,
            clock = clock,
            generatedIds = IdSource(ACK_A_ID)
        )
        first.send(
            ReliableChannelSendInputV1(
                ReliableFrameTypeV1.ROUTE_ADVERTISEMENT,
                "route-test".toByteArray(),
                durable = true,
                messageId = MESSAGE_ID
            )
        )
        first.close()
        assertEquals(1, durableStore.snapshot().outboxDepth)

        val outbound = mutableListOf<ByteArray>()
        val acknowledgements = mutableListOf<ByteArray>()
        val delivered = mutableListOf<ByteArray>()
        val receiverStore = InMemoryReliableChannelStoreV1()
        val restored = channel(
            transport = { outbound += it.copyOf() },
            store = durableStore,
            clock = clock,
            generatedIds = IdSource(ACK_A_ID)
        )
        val receiver = channel(
            transport = { acknowledgements += it.copyOf() },
            store = receiverStore,
            clock = clock,
            txKey = KEY_B,
            rxKey = KEY_A,
            txPrefix = PREFIX_B,
            rxPrefix = PREFIX_A,
            generatedIds = IdSource(ACK_B_ID),
            peerTrustId = PEER_TRUST_A,
            onMessage = { delivered += it.payload.copyOf() }
        )

        try {
            assertEquals(1, restored.restoreDurableOutbox())
            drain(outbound, receiver)
            drain(acknowledgements, restored)

            assertEquals(0, durableStore.snapshot().outboxDepth)
            assertEquals(0, restored.snapshot().pendingMessages)
            assertArrayEquals("route-test".toByteArray(), delivered.single())
        } finally {
            restored.close()
            receiver.close()
            durableStore.close()
            receiverStore.close()
            abandonedFrames.forEach { it.fill(0) }
            outbound.forEach { it.fill(0) }
            acknowledgements.forEach { it.fill(0) }
            delivered.forEach { it.fill(0) }
        }
    }

    @Test
    fun `retry schedule suspends durable messages and explicit resume restarts attempts`() {
        val clock = MutableClock(NOW)
        val frames = mutableListOf<ByteArray>()
        val store = InMemoryReliableChannelStoreV1()
        val channel = channel(
            transport = { frames += it.copyOf() },
            store = store,
            clock = clock,
            maxAttempts = 3,
            baseRetryMs = 100,
            generatedIds = IdSource(ACK_A_ID)
        )

        try {
            channel.send(
                ReliableChannelSendInputV1(
                    ReliableFrameTypeV1.DATA,
                    "test-only".toByteArray(),
                    durable = true,
                    messageId = MESSAGE_ID
                )
            )
            val initialFrameCount = frames.size

            clock.value = NOW + 100
            assertEquals(1, channel.tick().retried)
            clock.value = NOW + 300
            assertEquals(1, channel.tick().retried)
            clock.value = NOW + 700
            val suspended = channel.tick()
            assertEquals(1, suspended.suspended)
            assertEquals(1, channel.snapshot().suspendedMessages)
            assertEquals(1, store.snapshot().outboxDepth)
            assertEquals(2, channel.snapshot().retries)
            assertEquals(initialFrameCount * 3, frames.size)

            assertEquals(1, channel.resumeSuspended())
            assertEquals(1, channel.tick().retried)
            assertEquals(0, channel.snapshot().suspendedMessages)
            assertEquals(1, store.snapshot().outboxDepth)
        } finally {
            channel.close()
            store.close()
            frames.forEach { it.fill(0) }
        }
    }

    @Test
    fun `TTL expiry clears pending durable state`() {
        val clock = MutableClock(NOW)
        val store = InMemoryReliableChannelStoreV1()
        val channel = channel(
            transport = {},
            store = store,
            clock = clock,
            generatedIds = IdSource(ACK_A_ID)
        )

        try {
            channel.send(
                ReliableChannelSendInputV1(
                    ReliableFrameTypeV1.DATA,
                    byteArrayOf(1),
                    durable = true,
                    ttlMs = 1_000,
                    messageId = MESSAGE_ID
                )
            )
            clock.value = NOW + 1_000
            val result = channel.tick()

            assertEquals(1, result.expired)
            assertEquals(0, store.snapshot().outboxDepth)
            assertEquals(0, channel.snapshot().pendingMessages)
        } finally {
            channel.close()
            store.close()
        }
    }

    @Test
    fun `upper layer failure rolls back dedup and allows a clean retry`() {
        val clock = MutableClock(NOW)
        val senderFrames = mutableListOf<ByteArray>()
        var deliveryAttempts = 0
        val receiverStore = InMemoryReliableChannelStoreV1()
        val sender = channel(
            transport = { senderFrames += it.copyOf() },
            store = InMemoryReliableChannelStoreV1(),
            clock = clock,
            generatedIds = IdSource(ACK_A_ID)
        )
        val receiver = channel(
            transport = {},
            store = receiverStore,
            clock = clock,
            txKey = KEY_B,
            rxKey = KEY_A,
            txPrefix = PREFIX_B,
            rxPrefix = PREFIX_A,
            generatedIds = IdSource(ACK_B_ID),
            peerTrustId = PEER_TRUST_A,
            onMessage = {
                deliveryAttempts += 1
                throw IllegalStateException("diagnostic consumer unavailable")
            }
        )

        try {
            sender.send(
                ReliableChannelSendInputV1(
                    ReliableFrameTypeV1.DATA,
                    "test-only".toByteArray(),
                    messageId = MESSAGE_ID
                )
            )
            val original = senderFrames.map { it.copyOf() }
            assertChannelError("DELIVERY_FAILED") {
                original.forEach(receiver::receiveFragment)
            }
            assertEquals(0, receiverStore.snapshot().inboxDedupDepth)
            assertChannelError("DELIVERY_FAILED") {
                original.forEach(receiver::receiveFragment)
            }
            assertEquals(2, deliveryAttempts)
            assertEquals(2, receiver.snapshot().deliveryFailures)
        } finally {
            sender.close()
            receiver.close()
            receiverStore.close()
            senderFrames.forEach { it.fill(0) }
        }
    }

    @Test
    fun `durable records never restore across peer trust contexts`() {
        val clock = MutableClock(NOW)
        val store = InMemoryReliableChannelStoreV1()
        val first = channel(
            transport = {},
            store = store,
            clock = clock,
            peerTrustId = PEER_TRUST_A,
            generatedIds = IdSource(ACK_A_ID)
        )
        first.send(
            ReliableChannelSendInputV1(
                ReliableFrameTypeV1.DATA,
                "peer-a".toByteArray(),
                durable = true,
                messageId = MESSAGE_ID
            )
        )
        first.close()

        val otherPeer = channel(
            transport = {},
            store = store,
            clock = clock,
            peerTrustId = PEER_TRUST_B,
            generatedIds = IdSource(ACK_B_ID)
        )
        try {
            assertEquals(0, otherPeer.restoreDurableOutbox())
            otherPeer.send(
                ReliableChannelSendInputV1(
                    ReliableFrameTypeV1.DATA,
                    "peer-b".toByteArray(),
                    durable = true,
                    messageId = MESSAGE_ID
                )
            )
            assertEquals(2, store.snapshot().outboxDepth)
        } finally {
            otherPeer.close()
            store.close()
        }
    }

    @Test
    fun `clock regression invalid random and reserved ACK fail closed`() {
        val clock = MutableClock(NOW)
        val channel = channel(
            transport = {},
            store = InMemoryReliableChannelStoreV1(),
            clock = clock,
            random = { 1.0 },
            generatedIds = IdSource(ACK_A_ID)
        )

        try {
            assertChannelError("ACK_RESERVED") {
                channel.send(
                    ReliableChannelSendInputV1(
                        ReliableFrameTypeV1.ACK,
                        byteArrayOf(),
                        messageId = MESSAGE_ID
                    )
                )
            }
            assertChannelError("INVALID_RANDOM_SOURCE") {
                channel.send(
                    ReliableChannelSendInputV1(
                        ReliableFrameTypeV1.DATA,
                        byteArrayOf(),
                        messageId = MESSAGE_ID
                    )
                )
            }
            assertChannelError("OUTBOX_CONFLICT") {
                channel.send(
                    ReliableChannelSendInputV1(
                        ReliableFrameTypeV1.DATA,
                        byteArrayOf(),
                        messageId = MESSAGE_ID
                    )
                )
            }

            clock.value = NOW - 1
            assertChannelError("CLOCK_REGRESSION") { channel.tick() }
        } finally {
            channel.close()
        }
    }

    private fun channel(
        transport: ReliableChannelTransportV1,
        store: ReliableChannelStoreV1,
        clock: MutableClock,
        txKey: ByteArray = KEY_A,
        rxKey: ByteArray = KEY_B,
        txPrefix: ByteArray = PREFIX_A,
        rxPrefix: ByteArray = PREFIX_B,
        maxAttempts: Int = 5,
        baseRetryMs: Long = 250,
        random: () -> Double = { 0.0 },
        generatedIds: IdSource,
        peerTrustId: String = PEER_TRUST_B,
        onMessage: ReliableChannelMessageHandlerV1 = ReliableChannelMessageHandlerV1 {}
    ): ReliableChannelV1 =
        ReliableChannelV1(
            transport = transport,
            peerTrustId = peerTrustId,
            mtu = 64,
            txKey = txKey,
            rxKey = rxKey,
            txNoncePrefix = txPrefix,
            rxNoncePrefix = rxPrefix,
            onMessage = onMessage,
            store = store,
            maxAttempts = maxAttempts,
            baseRetryMs = baseRetryMs,
            maxRetryMs = 8_000,
            random = random,
            now = { clock.value },
            messageIdGenerator = generatedIds::next
        )

    private fun drain(
        frames: MutableList<ByteArray>,
        receiver: ReliableChannelV1
    ): List<ReliableChannelReceiveResultV1> {
        val copy = frames.toList()
        frames.clear()
        return copy.map(receiver::receiveFragment)
    }

    private fun assertChannelError(code: String, operation: () -> Unit) {
        val error = assertThrows(ReliableChannelException::class.java) { operation() }
        assertEquals(code, error.code)
    }

    private class MutableClock(var value: Long)

    private class IdSource(vararg values: String) {
        private val values = ArrayDeque(values.toList())

        fun next(): String = values.removeFirst()
    }

    companion object {
        private const val NOW = 1_800_000_000_000L
        private const val MESSAGE_ID = "00112233445566778899aabbccddeeff"
        private const val ACK_A_ID = "10112233445566778899aabbccddeeff"
        private const val ACK_B_ID = "20112233445566778899aabbccddeeff"
        private const val ACK_B_DUPLICATE_ID = "30112233445566778899aabbccddeeff"
        private const val PEER_TRUST_A =
            "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
        private const val PEER_TRUST_B =
            "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
        private val KEY_A = ByteArray(32) { 0x11 }
        private val KEY_B = ByteArray(32) { 0x22 }
        private val PREFIX_A = ByteArray(8) { 0x33 }
        private val PREFIX_B = ByteArray(8) { 0x44 }
    }
}
